### ISSUE TITLE
ZMQ and deepcopy: TypeError: context must be specified

### ISSUE THREAD

**Issue by tritemio** [ipython/ipython#8211]
Hi,

we encounter an error running  lmfit master on ipython notebook (2.3, 2.4 and 3.0). By reading the stack trace my wild guess is that we have a strange interaction between zmq and deepcopy. These tests were run on windows 7 x64 and anaconda. pyzmq version was always 14.5.0.

The trace ends with:

```
C:\Users\laser2002j\Anaconda\lib\copy_reg.pyc in __newobj__(cls, *args)
     91 
     92 def __newobj__(cls, *args):
---> 93     return cls.__new__(cls, *args)
     94 
     95 def _slotnames(cls):

zmq/backend/cython/socket.pyx in zmq.backend.cython.socket.Socket.__cinit__ (zmq\backend\cython\socket.c:2365)()

TypeError: context must be specified
```

Any idea what is going wrong?

This is the original lmfit issue with full trace: https://github.com/lmfit/lmfit-py/issues/221

An this is simple notebooks showing the problem:

- https://gist.github.com/yoavram/3890c2d169b591172ab2

**Comment by takluyver** [comment]
It looks like a ZMQ socket, or something which has a reference to one, is ending up as a parameter value in the lmfit model, and then deepcopy is failing. I don't think anything with a reference to a ZMQ socket should be a model parameter. Is it possible that something is scraping the global namespace to make parameters?

**Comment by tritemio** [comment]
It don't think there is any global name scraping (there shouldn't be). We now have a much simpler example that reproduces the error:

```
import lmfit
print('lmfit version: %s' % lmfit.__version__)
from copy import deepcopy
from lmfit import Parameters
pars = Parameters()
pars.add('x', 1.0)
c = deepcopy(pars)
```

The puzzling thing is that the above example works in python an ipython text terminal (with no GUI).

**Comment by takluyver** [comment]
That's not especially puzzling - neither of those have ZMQ sockets, so I wouldn't expect them to hit this bug.

It looks like the trouble is with asteval, and I do see some namespace scraping going on there:
https://github.com/lmfit/lmfit-py/blob/master/lmfit/asteval.py#L88

**Comment by takluyver** [comment]
Try sticking a `print(key)` in just before this line:

https://github.com/lmfit/lmfit-py/blob/master/lmfit/parameter.py#L48

That should indicate what's causing the problem.

**Comment by tritemio** [comment]
@takluyver, thanks, I'm not familiar with the asteval code, @newville may have more insights.

For the record, adding the print you suggested I get:

```
...
BytesWarning
spacing
print
```
 followed by the error.

**Comment by takluyver** [comment]
Looking at the code, print refers to [Interpreter._printer](https://github.com/lmfit/lmfit-py/blob/master/lmfit/asteval.py#L477). So by deep copying that, you will copy the entire `lmfit.asteval.Interpreter` instance, which has a reference to `sys.stdout` as `self.writer`. In IPython, `sys.stdout` is replaced with an object which sends output over ZMQ to the frontend.

I doubt lmfit intends to be deepcopying that much.

**Comment by newville** [comment]
@tritemio @takluyver I haven't looked at this in great detail, but I agree that the basic problem is zmq vs whatever asteval has in it symbol table.  It's certainly believable that a dumb "copy everything" is not actually needed. Then again, we absolutely did not ask for zmq to be involved in the process at all. 

In fact, it looks to me like zmq is just plain broken. Probably by design.  I sure we can make it so that lmfit and asteval work around this deep bug in IPython, but the problem is square and undeniably there, not with either asteval or lmfit. 

If a script works in Python and not IPython, the problem is clearly with IPython.   

Are you going to fix it, or do you expect a work-around from us?

**Comment by takluyver** [comment]
I don't think you can assume that deep copying arbitrary things (like sys.stdout, which it's common to replace in order to capture or redirect output) is safe. I'd be inclined to say this is a bug in lmfit, in that it shouldn't be trying to deep copy objects it doesn't know about. It's up to @minrk whether a fix in zmq is appropriate - I'm pretty sure there's nothing we can do in IPython.

> If a script works in Python and not IPython, the problem is clearly with IPython. 

I don't think that's necessarily true. The things that we do can expose bugs in other code that wouldn't otherwise show up. That doesn't happen very often, but we do see it. We've even uncovered one or two problems in Python itself, I think. If my website worked in one browser but not another, there's a good chance that it's my HTML at fault, not the browser.

**Comment by newville** [comment]
@takluyver   Well, we can almost certainly make lmfit work around this issue, but if zmq or IPython breaks `deepcopy(sys.stdout)` because it monkey-patched `sys.stdout`,  we're definitely working around a problem we did not create.   

That is, one *can* `deepcopy(sys.stdout)`.  If some library or application replaces `sys.stdout` with something than cannot be deepcopy()'ed, the replacement is flawed.

The analogy with HTML is not relevant. The code in question works in Python.  If using some additional library breaks that (because, perhaps, it has replaced a function in the `sys` module with something that does not support the functionality of the original), the library is at fault.  You can't just replace `sys.stdout` with `None` and then say that any script that expects `print()` to work is broken.

**Comment by minrk** [comment]
If anything, the problem is not in IPython, but zmq, a dependency of IPython.

**Comment by ssanderson** [comment]
FWIW, `sys.stdout` is **not** deepcopy-able in Python3.4:
```
Python 3.4.0 (default, Apr 11 2014, 13:05:11)
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sys import stdout
>>> from copy import deepcopy
>>> deepcopy(stdout)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ssanderson/.virtualenvs/research/lib/python3.4/copy.py", line 174, in deepcopy
    rv = reductor(2)
TypeError: cannot serialize '_io.TextIOWrapper' object
```
compare to Python 2.7, where it is deepcopyable:
```
Python 2.7.6 (default, Mar 22 2014, 22:59:56)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from copy import deepcopy
>>> from sys import stdout
>>> deepcopy(stdout)
<closed file '<uninitialized file>', mode '<uninitialized file>' at 0x7f2e40026540>
```

**Comment by takluyver** [comment]
Quite a few of the functions and objects in `sys` are intended to be set to other objects, and sys.stdout is absolutely something you should expect to be replaced. Test code does that all the time, and anything which wants output to go somewhere other than a terminal, like IPython or IDLE, will do it.

**Comment by ssanderson** [comment]
@takluyver I think the argument isn't that it's unreasonable to replace sys.stdout, it's that the object you replace sys.stdout with should meet the same interface as sys.stdout.  Whether or not you should consider being deepcopyable part of the public interface of sys.stdout is maybe debatable, but I think as @newville points out, it would justifiably be bug in IPython if you replaced stdout with `None`.

**Comment by takluyver** [comment]
If your code could run in a Windows GUI environment, it should be able to handle `sys.stdout` being None. That's admittedly unusual, but not unheard of.

**Comment by Carreau** [comment]
Devil advocate, we can still implement a `__depcopy__` on ZMQ stream could we ?

**Comment by minrk** [comment]
We could monkeypatch to get around bugs in dependencies, or we could just report bugs to those dependencies. Since I'm the maintainer of the upstream project, there's no reason for us to be doing any monkeypatching.

**Comment by ssanderson** [comment]
@minrk I opened https://github.com/zeromq/pyzmq/issues/661 and linked back here.

**Comment by minrk** [comment]
Fixed by zeromq/pyzmq#662
